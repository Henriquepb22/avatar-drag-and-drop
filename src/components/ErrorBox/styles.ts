import styled, { css } from 'styled-components'

import * as AvatarUpload from 'components/AvatarUpload/styles'
import * as Image from 'components/Image/styles'

export const ErrorBox = styled(AvatarUpload.Content)`
    ${Image.Image} {
        width: auto;
        height: auto;
    }
`

export const ErrorInfo = styled.div`
    display: flex;
    flex-direction: column;
`

export const ErrorMessage = styled.p`
    ${({ theme }) => css`
        color: ${theme.colors.red};
        font-size: ${theme.font.sizes.medium};
        letter-spacing: -0.02em;
        line-height: 180%;
    `}
`

export const ErrorButton = styled.button`
    ${({ theme }) => css`
        width: fit-content;
        font-family: ${theme.font.family};
        border: 0;
        background: none;
        line-height: 180%;
        color: ${theme.colors.primary};
        font-size: ${theme.font.sizes.medium};
        font-weight: ${theme.font.medium};
        letter-spacing: -0.02em;
        text-decoration-line: underline;

        &:hover {
            cursor: pointer;
        }
    `}
`
