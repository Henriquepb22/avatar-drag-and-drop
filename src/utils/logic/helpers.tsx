const isImageFile = (file: File) => {
    const fileType = file.type
    const baseMimeType = fileType.split('/')[0]

    if (baseMimeType !== 'image') return false
    return true
}

export const getFileDataUrl = async (file: File): Promise<string> => {
    const reader = new FileReader()

    return new Promise((resolve, reject) => {
        const isImageType = isImageFile(file)

        if (isImageType) {
            reader.readAsDataURL(file)
            reader.onload = () => {
                if (reader.result) {
                    resolve(reader.result.toString())
                }
            }
        } else {
            reader.abort()
            reject(new Error('File is not an recognized image type'))
        }
    })
}

export const cropImage = async (
    imageSrc: string,
    zoomLevel: number
): Promise<string> => {
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    const image = new Image()

    return new Promise((resolve, reject) => {
        image.src = imageSrc

        image.onload = () => {
            if (context) {
                const needCompression =
                    Math.max(canvas.width, canvas.height) > 1000

                canvas.width = needCompression ? image.width / 2 : image.width
                canvas.height = needCompression
                    ? image.height / 2
                    : image.height

                context.drawImage(image, 0, 0, canvas.width, canvas.height)
                context.globalCompositeOperation = 'destination-in'
                context.beginPath()
                context.arc(
                    canvas.width * 0.5,
                    canvas.height * 0.5,
                    (Math.max(canvas.width, canvas.height) * 0.5) /
                        (zoomLevel * 2),
                    0,
                    2 * Math.PI
                )
                context.fill()
                context.globalCompositeOperation = 'source-over'

                resolve(canvas.toDataURL('image/png', 1))
            }
        }

        image.onerror = () => {
            reject(new Error('Something bad happened...'))
        }
    })
}
