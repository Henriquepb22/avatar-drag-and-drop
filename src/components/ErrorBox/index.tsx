import * as S from './styles'

import infoIcon from 'assets/img/icons/info.svg'

type ErrorBoxProps = {
    handleReset: () => void
}

const ErrorBox = ({ handleReset }: ErrorBoxProps) => {
    return (
        <S.ErrorBox>
            <S.ImageContainer>
                <img src={infoIcon} alt="exclamation point" />
            </S.ImageContainer>
            <S.ErrorInfo>
                <S.ErrorMessage>Sorry, the upload failed.</S.ErrorMessage>
                <S.ErrorButton type="button" onClick={handleReset}>
                    Try again
                </S.ErrorButton>
            </S.ErrorInfo>
        </S.ErrorBox>
    )
}

export default ErrorBox
