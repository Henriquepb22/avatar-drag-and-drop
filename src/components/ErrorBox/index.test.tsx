import { renderWithTheme } from 'utils/tests/helpers'
import userEvent from '@testing-library/user-event'
import { screen } from '@testing-library/dom'

import ErrorBox from '.'

describe('<ErrorBox />', () => {
    it('should render error box correctly', () => {
        const handleReset = jest.fn()
        const { container } = renderWithTheme(
            <ErrorBox handleReset={handleReset} />
        )

        expect(
            screen.getByText(/sorry, the upload failed/i)
        ).toBeInTheDocument()
        expect(
            screen.getByRole('button', { name: /try again/i })
        ).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should call handleReset function on button click', () => {
        const handleReset = jest.fn()
        renderWithTheme(<ErrorBox handleReset={handleReset} />)

        const resetButton = screen.getByRole('button', { name: /try again/i })
        userEvent.click(resetButton)

        expect(handleReset).toHaveBeenCalledTimes(1)
    })
})
