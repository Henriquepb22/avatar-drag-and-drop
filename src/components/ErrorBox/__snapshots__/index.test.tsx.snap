// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBox /> should render error box correctly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #C3CBD5;
  border-radius: 50%;
  width: 11.4rem;
  height: 11.4rem;
  overflow: hidden;
}

.c4 {
  width: 22.8rem;
  height: 22.8rem;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  object-fit: cover;
}

.c0 {
  display: grid;
  grid-template-columns: 11.4rem 1fr;
  grid-gap: 3.2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .c3 {
  width: auto;
  height: auto;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  color: #C64D32;
  font-size: 1.6rem;
  -webkit-letter-spacing: -0.02em;
  -moz-letter-spacing: -0.02em;
  -ms-letter-spacing: -0.02em;
  letter-spacing: -0.02em;
  line-height: 180%;
}

.c7 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  font-family: Inter,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  background: none;
  line-height: 180%;
  color: #3D485F;
  font-size: 1.6rem;
  font-weight: 500;
  -webkit-letter-spacing: -0.02em;
  -moz-letter-spacing: -0.02em;
  -ms-letter-spacing: -0.02em;
  letter-spacing: -0.02em;
  -webkit-text-decoration-line: underline;
  text-decoration-line: underline;
}

.c7:hover {
  cursor: pointer;
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  >
    <img
      alt="exclamation point"
      class="c3 c4"
      src="info.svg"
    />
  </div>
  <div
    class="c5"
  >
    <p
      class="c6"
    >
      Sorry, the upload failed.
    </p>
    <button
      class="c7"
      type="button"
    >
      Try again
    </button>
  </div>
</div>
`;
